### PR TITLE
[#176] Adding Ledger support for transfer transactions

### LINKED ISSUE #176: Integrate Ledger functionality into LTO Mainnet Wallet
Ledger LTO App has reached developer mode status. LTO Mainnet wallet will need to be adjusted to integrate the Ledger device functionality.

Ledger App on Ignacio's github: https://github.com/iicc1/ledger-app-lto
The same repo, but forked by Ledger: https://github.com/LedgerHQ/app-lto

Wallet UI as on lto.stakely.io (third party ledger integrated wallet UI): https://github.com/iicc1/lto-network-ledger-wallet-ui

To currently find LTO App in Ledger Live Manager you'll have to turn on developer mode:
developermode

nanos

### PR DESCRIPTION
**Pull request by brunodb3**
- Added integration with Ledger devices via the [LTO Ledger App](https://github.com/iicc1/ledger-app-lto)
- Added necessary libraries to make that happen
- Only `transfer` transactions are available, as further development needs to be done on the [Ledger app](https://github.com/iicc1/ledger-app-lto) to support other transaction types (see [issue #4](https://github.com/iicc1/ledger-app-lto/issues/4))
  - Because of that, a `popup` has been added to the pages of the transaction types that are not supported by Ledger, so that the user knows about it
- Added script for `server`, where the `bridge` is mocked (data is on `/mock-data/lto-bridge.json`). `json-server` is used to supply an endpoint on `localhost` with the mocked data from the Bridge, since testnet bridge is not reliable (there are CORS issues, as well as it's not maintained).
- Under `polyfills.ts` some workarounds were added, result of adding the Ledger-related libraries as dependencies
- Some minor fixes and formatting

--- 

closes #176

### CODE REVIEW

File: apps/wallet/src/app/core/services/ledger.service.ts [hunk 1]
```diff
@@ -0,0 +1,286 @@
...
+
+import { BehaviorSubject } from 'rxjs';
+import TransportU2F from '@ledgerhq/hw-transport-u2f';
+import TransportWebUSB from '@ledgerhq/hw-transport-webusb';
+import { WavesLedger } from 'lto-ledger-js-unofficial-test';
+
+import { TransactionTypes } from '@wallet/core/transaction-types';
+import { LTO_NETWORK_BYTE, LTO_PUBLIC_API } from '@wallet/tokens';
+
+import {
```
Comment (KaloyanTanev) [inline apps/wallet/src/app/core/services/ledger.service.ts:12]: sponsor as well?

File: apps/wallet/src/app/core/services/ledger.service.ts [hunk 2]
```diff
@@ -0,0 +1,286 @@
...
+          attachment: data.attachment,
+          type: data.type as unknown as TRANSACTION_TYPE.TRANSFER,
+        };
+
+        unsignedTransaction = transfer(rawTransaction);
+
+        prefixBytes = new Uint8Array();
+        schema = parseSerialize.schemas.transferSchemaV1;
+        break;
+      case TransactionTypes.MASS_TRANSFER:
```
Comment (KaloyanTanev) [inline apps/wallet/src/app/core/services/ledger.service.ts:150]: If those won't be supported, I suggest comment them out (security concerns)?

File: apps/wallet/src/environments/environment.local.ts [hunk 3]
```diff
@@ -0,0 +1,11 @@
...
+  production: true,
+  networkByte: 'T',
+  transfer_fee: 100000000,
+  mass_transfer_fee: 10000000,
+  anchor_fee: 35000000,
+  bridge: true,
+  swapPageEnabled: false,
+  lto_api_url: 'https://testnet.lto.network/',
+  bridge_url: 'http://localhost:3000'
+};
```
Comment (KaloyanTanev) [inline apps/wallet/src/environments/environment.local.ts:11]: New line at EOF

File: apps/wallet/src/polyfills.ts [hunk 4]
```diff
@@ -78,3 +78,9 @@ import 'zone.js/dist/zone';  // Included with Angular CLI.
 /***************************************************************************************************
  * APPLICATION IMPORTS
  */
+// https://stackoverflow.com/questions/54349858/angular-7-uncaught-referenceerror-global-is-not-defined-when-adding-package
+(window as any)['global'] = window;
+// https://stackoverflow.com/questions/50371593/angular-6-uncaught-referenceerror-buffer-is-not-defined
+global.Buffer = global.Buffer || require('buffer').Buffer;
+// https://stackoverflow.com/questions/50313745/angular-6-process-is-not-defined-when-trying-to-serve-application
+(window as any).process = { env: { DEBUG: undefined } };
```
Comment (KaloyanTanev) [inline apps/wallet/src/polyfills.ts:86]: New line at EOF

File: mock-data/lto-bridge.json [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+      "binance": {
+        "address": "bnb1esdfkkddv0sa5w7njk080rwuda47k3me3nxk75",
+        "total": 500000000,
+        "bridge": 467664120.75583875,
+        "supply": 32335879.24416125,
+        "burn_fee": 1000000000
+      }
+    }
+  }
+}
```
Comment (KaloyanTanev) [inline mock-data/lto-bridge.json:32]: New line at EOF

File: tsconfig.json [hunk 6]
```diff
@@ -14,38 +14,14 @@
     "target": "es2015",
     "module": "es2020",
     "lib": ["es2018", "dom"],
+    "skipLibCheck": true,
```
Comment (KaloyanTanev) [inline tsconfig.json:17]: What does this one do?

File: apps/wallet/src/app/components/content-dialog/content-dialog.component.html [hunk 7]
```diff
@@ -0,0 +1,4 @@
+<h2 mat-dialog-title>{{ data.title }}</h2>
+<mat-dialog-content>
+  <p>{{ data.content }}</p>
+</mat-dialog-content>
```
[the remark this answers is not in the document]
Comment (brunodb3) [reply apps/wallet/src/app/components/content-dialog/content-dialog.component.html:4]: It's just for keeping the standard really, every component has a styling file attached to it. Could remove, but it does no harm and it's easier to change something if needed.

File: apps/wallet/src/app/core/services/ledger.service.ts (hunk 1 above)
Comment (brunodb3) [reply apps/wallet/src/app/core/services/ledger.service.ts:12]: AFAIK the wallet doesn't support sponsor transactions (yet at least), neither does the Ledger app

File: apps/wallet/src/app/core/services/ledger.service.ts (hunk 2 above)
Comment (brunodb3) [reply apps/wallet/src/app/core/services/ledger.service.ts:150]: What are the security concerns?

File: tsconfig.json (hunk 6 above)
Comment (brunodb3) [reply tsconfig.json:17]: Skips checking of TypeScript for libraries inside `node_modules`. Had to add it after installing libs like `lto-transactions` that require a different TS version, so it gives out warnings. Ideally the lib would update to same TS version, but as long as lib works, I don't see an issue

File: apps/wallet/src/polyfills.ts (hunk 4 above)
Comment (brunodb3) [reply apps/wallet/src/polyfills.ts]: Done

File: apps/wallet/src/environments/environment.local.ts (hunk 3 above)
Comment (brunodb3) [reply apps/wallet/src/environments/environment.local.ts]: Done

File: mock-data/lto-bridge.json (hunk 5 above)
Comment (brunodb3) [reply mock-data/lto-bridge.json]: Done

**Review by jasny — APPROVED** [review]

File: apps/wallet/src/app/core/services/ledger.service.ts (hunk 2 above)
Comment (KaloyanTanev) [reply apps/wallet/src/app/core/services/ledger.service.ts:150]: Can this method be accessed from an API call? If so when someone makes tx with such type, we should output the default case: `throw new Error('Unknown transaction type')` (or even better - 'Not supported transaction type') and not put it down the line and mess up things eventually?

**Pull request merged**